packages/medusa/src/services/order.ts

```diff
@@ -1185,6 +1193,108 @@ class OrderService extends TransactionBaseService {
     })
   }
 
+  async confirmCustomerClaimToOrders(token: string, customerId: string) {
+    return await this.atomicPhase_(async (manager) => {
+      logger.info(token)
+      const { claimingCustomerId, orders: orderIds } = this.verifyToken(
+        token
+      ) as {
+        claimingCustomerId: string
+        orders: string[]
+      }
+
+      if (customerId !== claimingCustomerId) {
+        throw new MedusaError(
+          MedusaError.Types.UNAUTHORIZED,
+          `The token is not valid`
+        )
+      }
+
+      const customer = await this.customerService_
+        .withTransaction(manager)
+        .retrieve(customerId)
+
+      const orderRepo = manager.getCustomRepository(this.orderRepository_)
+
+      const orders = await orderRepo.findByIds(orderIds)
+
+      for (const order of orders) {
+        order.customer_id = claimingCustomerId
+        order.email = customer.email
+      }
+
+      await orderRepo.save(orders)
+    })
+  }
+
+  async claimOrdersForCustomerWithId(
+    customerId: string,
+    orderIds: string[]
+  ): Promise<void> {
+    return await this.atomicPhase_(async (manager) => {
+      const customer = await this.customerService_
+        .withTransaction(manager)
+        .retrieve(customerId)
+
+      if (!customer.has_account) {
+        throw new MedusaError(
+          MedusaError.Types.UNAUTHORIZED,
+          "Customer does not have an account"
+        )
+      }
+
+      const orderRepo = manager.getCustomRepository(this.orderRepository_)
+      const orders = await orderRepo.findByIds(orderIds)
+
+      const emailOrderMapping: { [email: string]: Order[] } = orders.reduce(
+        (acc, order) => {
+          acc[order.email] = [...(acc[order.email] || []), order]
+          return acc
+        },
+        {}
+      )
+
+      await Promise.all(
+        Object.keys(emailOrderMapping).map(async (email) => {
+          const token = this.signToken({
+            claimingCustomerId: customerId,
+            orders: emailOrderMapping[email].map((o) => o.id),
+          })
+
+          await this.eventBus_.emit(OrderService.Events.ORDERS_CLAIMED, {
+            email,
+            claimingCustomerEmail: customer.email,
+            orders: emailOrderMapping[email],
+            token,
+          })
+        })
```

Comment:
thought: we can open this up a bit to enable the subscriber (often a notification plugin) to collect the data it needs.

```suggestion
          await this.eventBus_.emit(OrderService.Events.ORDERS_CLAIMED, {
            old_email: email,
            new_customer_id: customer.id,
            order_ids: emailOrderMapping[email].map((o) => o.id),
            token,
          })
        })
```